lib/template-loader.php

```diff
@@ -0,0 +1,162 @@
+<?php
+/**
+ * Block template loader functions.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Adds necessary filters to use 'wp_template' posts instead of theme template files.
+ */
+function gutenberg_add_template_loader_filters() {
+	if ( ! post_type_exists( 'wp_template' ) ) {
+		return;
+	}
+
+	/**
+	 * Array of all overrideable default template types.
+	 *
+	 * @see get_query_template
+	 *
+	 * @var array
+	 */
+	$template_types = array(
+		'index',
+		'404',
+		'archive',
+		'author',
+		'category',
+		'tag',
+		'taxonomy',
+		'date',
+		// Skip 'embed' for now because it is not a regular template type.
+		'home',
+		'frontpage',
+		'privacypolicy',
+		'page',
+		'search',
+		'single',
+		'singular',
+		'attachment',
+	);
+	foreach ( $template_types as $template_type ) {
+		add_filter( $template_type . '_template', 'gutenberg_override_query_template', 20, 3 );
+	}
+
+	add_filter( 'template_include', 'gutenberg_find_template', 20 );
+}
+add_action( 'wp_loaded', 'gutenberg_add_template_loader_filters' );
+
+/**
+ * Filters into the "{$type}_template" hooks to record the current template hierarchy.
+ *
+ * The method returns an empty result for every template so that a 'wp_template' post
+ * is used instead.
+ *
+ * @see gutenberg_find_template
+ *
+ * @param string $template  Path to the template. See locate_template().
+ * @param string $type      Sanitized filename without extension.
+ * @param array  $templates A list of template candidates, in descending order of priority.
+ * @return string Empty string to ensure template file is considered not found.
+ */
+function gutenberg_override_query_template( $template, $type, array $templates = array() ) {
+	global $_wp_current_template_hierarchy;
+
+	if ( ! is_array( $_wp_current_template_hierarchy ) ) {
+		$_wp_current_template_hierarchy = $templates;
+	} else {
+		$_wp_current_template_hierarchy = array_merge( $_wp_current_template_hierarchy, $templates );
```

Comment:
Why do we merge here? Merging two ordered lists like this doesn't seem like it would result in the right ordering.

Reply:
How so? As these are regular indexed arrays, the `$templates` one would always be appended to the `$_wp_current_template_hierarchy` one.

Reply:
Exactly, I'm not even sure if this runs more than once, but the code looks like it expects it too and if on every run `$templates` contains a priority ordered list, what guarantees that the head of the next list is the tail of the previous list?